tsdb/engine/tsm1/reader.go

```diff
@@ -560,6 +571,62 @@ func (t *TSMReader) TombstoneRange(key []byte) []TimeRange {
 	return tr
 }
 
+// BlockCount returns number of values stored in the block at location idx
+// in the file at path.  If path does not match any file in the store, 0 is
+// returned.  If idx is out of range for the number of blocks in the file,
+// 0 is returned.
+func (t *TSMReader) BlockCount(idx int) (int, error) {
+	if idx < 0 {
+		return 0, fmt.Errorf("%w: TSMReaderBlockCount: invalid index %d for %q", ErrBlockOutOfRange, idx, t.Path())
+	}
+
+	iter := t.BlockIterator()
+	for i := 0; i < idx; i++ {
+		if !iter.Next() {
+			return 0, fmt.Errorf("%w: TSMReader.BlockCount: error seeking to block %d for %q", ErrBlockOutOfRange, idx, t.Path())
+		}
+	}
+	_, _, _, _, _, block, err := iter.Read()
+	if err != nil {
+		return 0, fmt.Errorf("%w: TSMReaderBlockCount: error reading block %d in %q", err, idx, t.Path())
+	}
+	count, err := BlockCount(block)
+	if err != nil {
+		return 0, fmt.Errorf("%w: TSMReaderBlockCount: error finding block count for block %d in %q", err, idx, t.Path())
+	}
+	return count, nil
+}
+
+// firstBlockCount returns the number of points in the first block of the file.
+// firstBlockCount is determined on the first call and then a cached value is returned.
+// firstBlockCount should not be called with t.mu held.
+func (t *TSMReader) firstBlockCount() (int, error) {
+	cache := func() blockCountCache {
+		t.mu.RLock()
+		defer t.mu.RUnlock()
+		return t.firstBlockCountCache
+	}()
+
+	if !cache.valid {
+		count, err := t.BlockCount(1)
```

Comment:
@davidby-influx @devanbenz Currently `firstBlockCount` can not hold a lock on `t.mu` because the code would deadlock when it called `t.BlockCount`. I'm pretty sure that it is completely safe to call `t.BlockCount` without a lock on `t.mu`, but please verify.